platform/linux-generic/odp_crypto_ipsecmb.c

```diff
@@ -0,0 +1,999 @@
+
+		data = local.buffer;
+	}
+
+	if (session->p.auth_key.length == 16) {
+		/* ZUC128 EIA3 */
+                IMB_ZUC_EIA3_1_BUFFER(mb_mgr, (const void  **)session->auth.key,
+				      (const void *)iv_ptr,
+				      (const void *)data,
+                                      (param->auth_range.length) * 8,
+				      (void *)dst_data);
+	} else if (session->p.auth_key.length == 32) {
+		/* ZUC256 EIA3 */
+                IMB_ZUC256_EIA3_1_BUFFER(mb_mgr, (const void **)session->auth.key,
+					 (const void *)iv_ptr,
+					 (const void *)data,
+                                         (param->auth_range.length) * 8,
+					 (void *)dst_data);
+	}
+
+	/* Copy to the output location */
+	odp_packet_copy_from_mem(pkt, param->hash_result_offset,
+				 session->p.auth_digest_len,
+				 dst_data);
+
+	set_crypto_op_result_ok(pkt);
+}
+
+static
+void auth_zuc_eia3_check(odp_packet_t pkt,
+			 const odp_crypto_packet_op_param_t *param,
+			 odp_crypto_generic_session_t *session)
+{
+	IMB_MGR *mb_mgr = local.mb_mgr;
+	uint8_t *iv_ptr;
+	uint32_t in_pos = param->auth_range.offset;
+	uint32_t in_len = param->auth_range.length;
+	uint8_t  hash_in[IPSEC_MB_CRYPTO_MAX_DIGEST_LENGTH];
+        uint8_t dst_data[MAX_BUFFER_LENGTH_IN_BYTES];
+
+	/* Copy current value out and clear it before authentication */
+	odp_packet_copy_to_mem(pkt, param->hash_result_offset,
+			       session->p.auth_digest_len, hash_in);
+
+	_odp_packet_set_data(pkt, param->hash_result_offset,
+			     0, session->p.auth_digest_len);
+
+#if ODP_DEPRECATED_API
+	if (param->cipher_iv_ptr)
+		iv_ptr = param->auth_iv_ptr;
+	else if (session->p.auth_iv.data)
+		iv_ptr = session->auth.iv_data;
+	else
+		goto err;
+#else
+	iv_ptr = param->auth_iv_ptr;
+	_ODP_ASSERT(session->p.auth_iv_len == 0 || iv_ptr != NULL);
+#endif
+
+	uint32_t seg_len = 0;
+	uint8_t *data = odp_packet_offset(pkt, in_pos, &seg_len, NULL);
+
+	if (odp_unlikely(odp_packet_is_segmented(pkt))) {
+		/* Packet is segmented within the cipher range. Copy the cipher
+		 * range to a contiguous buffer. */
+		odp_packet_copy_to_mem(pkt, in_pos, in_len, local.buffer);
+
+		data = local.buffer;
+	}
+
+	if (session->p.auth_key.length == 16) {
+		/* ZUC128 EIA3 */
+                IMB_ZUC_EIA3_1_BUFFER(mb_mgr, (const void  **)session->auth.key,
+				      (const void *)iv_ptr,
+				      (const void *)data,
+                                      (param->auth_range.length) * 8,
+				      (void *)dst_data);
+	} else if (session->p.auth_key.length == 32) {
+		/* ZUC256 EIA3 */
+                IMB_ZUC256_EIA3_1_BUFFER(mb_mgr, (const void  **)session->auth.key,
+				      (const void *)iv_ptr,
+				      (const void *)data,
+                                      (param->auth_range.length) * 8,
+				      (void *)dst_data);
+        }
+
+	/* Verify match */
+	if (0 != memcmp(hash_in, dst_data, session->p.auth_digest_len))
+		goto err;
+
+	/* Copy to the output location */
+	odp_packet_copy_from_mem(pkt, param->hash_result_offset,
```

Comment:
According to the crypto API, the hash result bytes are left undefined after a decode operation, so this copying is not needed.

Reply:
Removed.